umbraco/Umbraco-CMS
# Fixing macro rendering with forms

Macros with forms were not working because none of current ViewData/ModelState/TempData were being carried through to the view context used to render macros.

## Main changes

* ProxyViewDataFeature now carries both ViewData and TempData so that we can use this during routing from a SurfaceController. Previously it was only carrying through ViewData
* Fixes PartialViewMacroEngine and ensure ViewData and TempData are carried across
* Fixes the ValidationSummary ext method to correctly deal with the validation ModelState prefixes

## Other

* new `UmbracoRoutedRequest` event - raised when we have a front-end routed request with an assigned content item. We already have `UmbracoRequestBegin` (which fires before routing takes place) and `UmbracoRequestEnd` which fires at the very end of the request. There was no event to know when a front-end request was attempted.
* We now use this event to ensure the `IUmbracoWebsiteSecurityAccessor` - but have added notes to this since we should kill this. The `IUmbracoWebsiteSecurity` is also just throwing exceptions since it's not complete. This is part of the members migration.
* Changes HybridAccessorBase to automatically assign a key instead of hard coding keys (these were out of date with the namespaces changes)
* Adds notes about Temporal Coupling with the IBackOfficeSecurity and UmbracoContext, we'll need to decouple these things entirely and not have one relying on transitive dependencies of the other

## Testing

* Create a new partial view macro from a snippet: use the login snippet
* Allow this macro to be used within an RTE
* Add this macro to an RTE
* Render the value in a view
* Modify your partial view .cshtml, add this somewhere to render out the TempData value when its successful. Currently the UmbLoginController just sets this to `true` (not actually sure why)
   ```cs
   if (TempData.TryGetValue("LoginSuccess", out var msg))
   {
       <div>@msg</div>
   }
   ```
* Add breakpoint to UmbLoginController since `_websiteSecurityAccessor.WebsiteSecurity.LoginAsync` is just going to throw... so you can manually change how the result works, either move the execution to the `AddModelError` to add an error and return once that is tested, test moving the execution to `TempData["LoginSuccess"] = true;` for a successful test. 
* You should see that the validation summary and TempData behave correctly

Bonus points:

* Create a custom ViewComponent that looks up the IPublishedContent item above which renders the RTE
* Then render this ViewComponent in a view of a different IPublishedContent
* Now we have multiple levels of ViewComponents and controllers
* The ViewData/ModelState/TempData should still work as expected
